apps/cowswap-frontend/src/modules/permit/utils/generatePermitHook.ts

```diff
@@ -20,86 +20,96 @@ function getCacheKey(params: PermitHookParams): string {
-            ],
-          })
-        : buildDaiLikePermitCallData({
-            eip2162Utils: eip2612PermitUtils,
-            callDataParams: [
-              {
-                holder: owner,
-                spender,
-                allowed: true,
-                value,
-                nonce,
-                expiry: deadline,
-              },
-              chainId as number,
-              tokenName,
-              tokenAddress,
-            ],
-          })
+  REQUESTS_CACHE[permitKey] = request
 
-    try {
-      const callData = await callDataPromise
+  return request
+}
 
-      const permitHookData = {
-        target: tokenAddress,
-        callData,
-        gasLimit: permitInfo.gasLimit.toString(),
-      }
+async function generatePermitHookRaw(params: PermitHookParams): Promise<PermitHookData> {
+  const { inputToken, chainId, permitInfo, provider, account } = params
+  const tokenAddress = inputToken.address
+  const tokenName = inputToken.name || tokenAddress
 
-      const permitHook = JSON.stringify(permitHookData)
+  // TODO: verify whether cached result is still valid and renew it if needed
 
-      localStorage.setItem(permitKey, permitHook)
+  const web3ProviderConnector = new PermitProviderConnector(provider, account ? undefined : PERMIT_SIGNER)
+  const eip2612PermitUtils = new Eip2612PermitUtils(web3ProviderConnector)
 
-      resolve(permitHookData)
-    } catch (e) {
-      return Promise.reject(e.message)
-    }
-  })
+  const owner = account || PERMIT_SIGNER.address
 
-  REQUESTS_CACHE[permitKey] = request
+  // TODO: check whether cached permit nonce matches current nonce and update it in case it doesnt
 
-  return request
+  const nonce = await eip2612PermitUtils.getTokenNonce(tokenAddress, owner)
+
+  const spender = GP_VAULT_RELAYER[chainId]
+  const deadline = getPermitDeadline()
+  const value = DEFAULT_PERMIT_VALUE
+
+  const callDataPromise =
+    permitInfo.type === 'eip-2612'
+      ? buildEip2162PermitCallData({
+          eip2162Utils: eip2612PermitUtils,
+          callDataParams: [
+            {
+              owner,
+              spender,
+              value,
+              nonce,
+              deadline,
+            },
+            chainId as number,
+            tokenName,
+            tokenAddress,
+          ],
+        })
+      : buildDaiLikePermitCallData({
+          eip2162Utils: eip2612PermitUtils,
+          callDataParams: [
+            {
+              holder: owner,
+              spender,
+              allowed: true,
+              value,
+              nonce,
+              expiry: deadline,
+            },
+            chainId as number,
+            tokenName,
+            tokenAddress,
+          ],
+        })
+
+  const callData = await callDataPromise
+
+  return {
+    target: tokenAddress,
+    callData,
+    // TODO: Find a way to calculate the gas for quote calls without requesting user to sign
+    // For now, keep it hard coded
+    gasLimit: '80000',
```

Comment:
Dumb approach to have enough room to account for variations in the gas limit

Otherwise we need to request the user's signature before sending a quote request, which is a bad UX